// aot/dist/src/AOT_Placeholder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AOT_FLAGS_CACHE = new WeakMap();
class AOT_Placeholder {
    static GetAOTFlags(target) {
        const aot_flags = [];
        let proto = target;
        do {
            const _flags = AOT_FLAGS_CACHE.get(proto);
            if (_flags) {
                aot_flags.push(..._flags);
            }
            proto = Object.getPrototypeOf(proto);
            if (!proto) {
                break;
            }
        } while (true);
        return aot_flags;
    }
    static GetAndSetAOTFlags(target) {
        let aot_flags = AOT_FLAGS_CACHE.get(target);
        if (!aot_flags) {
            aot_flags = [];
            AOT_FLAGS_CACHE.set(target, aot_flags);
        }
        return aot_flags;
    }
    static Then(then_fun_name) {
        return (target, name, des) => {
            this.GetAndSetAOTFlags(target).push({
                type: "Then",
                data: { then_fun_name, prop_name: name },
            });
            return des;
        };
    }
    static Wait(condition_promise_fun_name, skip_if_false = false) {
        return (target, name, des) => {
            this.GetAndSetAOTFlags(target).push({
                type: "Wait",
                data: { condition_promise_fun_name, skip_if_false, prop_name: name },
            });
            return des;
        };
    }
}
exports.AOT_Placeholder = AOT_Placeholder;
